### PR TITLE
feat: Add new techdocs-rewrite-relative-links-action

### PR DESCRIPTION
**Pull request by zerok**
This should help improve our internal techdoc generation and will most likely be part of the publish-techdocs workflow.

### CODE REVIEW

**Comment by zerok** [comment]
The diff in https://github.com/grafana/deployment_tools/actions/runs/8734877664/job/23973857969 looks actually quite sane! 🙂 

- emph with `_` is replaced with `*` (due to markdown re-rendering)
- indentation is normalized incl. line endings (due to markdown re-rendering)
- relative links to files point to `blob`
- relative links to folders point to `tree`

File: actions/techdocs-rewrite-relative-links/main.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			ctrl := controller{
+				logger:        logger,
+				rootDirectory: rootDir,
+				repoURL:       repoURL,
+				defaultBranch: defaultBranch,
+			}
+			return ctrl.run(ctx)
+		},
+	}
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/main.go:58]: Hah! I actually wanted to do that and then completely forgot about it 🤣

File: .github/workflows/publish-techdocs.yaml [hunk 2]
```diff
@@ -35,6 +43,23 @@ jobs:
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         if: inputs.checkout-repo
 
+      - name: Rewrite relative links
+        if: inputs.rewrite-relative-links
+        # TODO: Pin this action once stable
+        # Using an absolute reference to the action here due to the complexity
+        # involved with a composite action that also checks out its own code:
```
[the remark this answers is not in the document]
Comment (zerok) [reply .github/workflows/publish-techdocs.yaml:50]: It's mostly that the workflow needs to checkout the shared-workflows repo and then the action itself does the same again. Will try to get that working, though (and in a readable way) 🙂

File: actions/techdocs-rewrite-relative-links/main_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+			defaultBranch: "main",
+			repoURL:       "https://github.com/grafana/dummy",
+		}
+		docsDir, err := ctrl.getDocsDir(context.Background(), rootDir)
+		require.NoError(t, err)
+		require.Equal(t, filepath.Join("/tmp", "docs"), docsDir)
+	})
+}
+
+func createTestDirectory(t *testing.T, rootDir string, content map[string]string) {
```
[the remark this answers is not in the document]
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/main_test.go:101]: Unfortunately no, io/fs is read-only at this point and we need to update files. We could go with https://github.com/spf13/afero here, though.

File: .github/workflows/publish-techdocs.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (zerok) [reply .github/workflows/publish-techdocs.yaml]: Seems to work now 🙂

File: actions/techdocs-rewrite-relative-links/main_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/main_test.go]: Move to afero complete :)

File: actions/techdocs-rewrite-relative-links/README.md
[the remark this answers is not in the document]
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/README.md:1]: Done :)

File: .github/workflows/publish-techdocs.yaml [hunk 4]
```diff
@@ -35,6 +48,25 @@ jobs:
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         if: inputs.checkout-repo
 
+      - id: checkout-shared-workflows
+        name: Checkout shared workflows
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          repository: grafana/shared-workflows
+          ref: zerok/techdocs-rewrite-relative-links-action
```
Comment (zerok) [inline .github/workflows/publish-techdocs.yaml:56]: Update to main before merge and then to the latest commit.

File: actions/techdocs-rewrite-relative-links/action.yaml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  checkout-path:
+    type: string
+    default: _action
+    required: false
+  skip-checkout:
+    description: |
+      If the action is already present in the workspace you can skip another checkout.
+    type: boolean
+    required: false
+    default: false
```
Comment (iainlane) [inline actions/techdocs-rewrite-relative-links/action.yaml:29]: suggestion: Could be (probably is) just me, but I find inverted options hard to think about - I'm unsure whether to use `true` or `false`. Suggest calling this `checkout` instead and then it'd look like `checkout: false` to skip, which is nice enough I think. :slightly_smiling_face:

File: actions/techdocs-rewrite-relative-links/action.yaml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  checkout-path:
+    type: string
+    default: _action
+    required: false
+  skip-checkout:
+    description: |
+      If the action is already present in the workspace you can skip another checkout.
```
Comment (iainlane) [inline actions/techdocs-rewrite-relative-links/action.yaml:26]: might need to remind people that they probably need to set `checkout-path` if doing the checkout themselves

**Review by iainlane — APPROVED** [review]
Nice :+1: TIL about `afero`. Couple of suggestions but let's give this a go, thanks!

Suggest rebasing on (or reviewing) #108 first so that we don't get more drift :grin:

**Comment by zerok** [comment]
> Nice 👍 TIL about `afero`. Couple of suggestions but let's give this a go, thanks!
> 
> ~Suggest rebasing on (or reviewing) #108 first so that we don't get more drift 😁~ edit:merged now

Done :)

File: actions/techdocs-rewrite-relative-links/action.yaml (hunk 6 above)
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/action.yaml]: Done

File: actions/techdocs-rewrite-relative-links/action.yaml (hunk 5 above)
Comment (zerok) [reply actions/techdocs-rewrite-relative-links/action.yaml]: Done

**Pull request merged**